ouroboros-network/src/Ouroboros/Network/Codec.hs

```diff
@@ -40,36 +48,113 @@ cborCodec
+       -> DecodeStep BS.ByteString CBOR.DeserialiseFailure m a
     go (CBOR.Done  trailing _ x)
       | BS.null trailing       = DecodeDone x Nothing
       | otherwise              = DecodeDone x (Just trailing)
     go (CBOR.Fail _ _ failure) = DecodeFail failure
     go (CBOR.Partial k)        = DecodePartial (fmap go . liftST . k)
 
+
+-- | Construct a 'Codec' for a CBOR based serialisation format, using lazy
+-- 'BS.ByteString's.
+--
+-- This is an adaptor between the @cborg@ library and the 'Codec' abstraction.
+--
+-- It takes encode and decode functions for the protocol messages that use the
+-- CBOR library encoder and decoder.
+--
+mkCodecCborLazyBS
+  :: forall ps m. MonadST m
+
+  => (forall (pr :: PeerRole) (st :: ps) (st' :: ps).
+             PeerHasAgency pr st
+          -> Message ps st st' -> CBOR.Encoding)
+
+  -> (forall (pr :: PeerRole) (st :: ps) s.
+             PeerHasAgency pr st
+          -> CBOR.Decoder s (SomeMessage st))
+
+  -> Codec ps CBOR.DeserialiseFailure m LBS.ByteString
+mkCodecCborLazyBS  cborMsgEncode cborMsgDecode =
+    Codec {
+      encode = \stok msg -> convertCborEncoder (cborMsgEncode stok) msg,
+      decode = \stok     -> convertCborDecoder (cborMsgDecode stok)
+    }
+  where
+    convertCborEncoder :: (a -> CBOR.Encoding) -> a -> LBS.ByteString
+    convertCborEncoder cborEncode =
+        toLazyByteString
+      . CBOR.toBuilder
+      . cborEncode
+
+    convertCborDecoder
+      :: (forall s. CBOR.Decoder s a)
+      -> m (DecodeStep LBS.ByteString CBOR.DeserialiseFailure m a)
+    convertCborDecoder cborDecode =
+        withLiftST (convertCborDecoderLBS cborDecode)
+
+convertCborDecoderLBS
+  :: forall s m a. Monad m
+  => (CBOR.Decoder s a)
+  -> (forall b. ST s b -> m b)
+  -> m (DecodeStep LBS.ByteString CBOR.DeserialiseFailure m a)
+convertCborDecoderLBS cborDecode liftST =
+    go [] =<< liftST (CBOR.deserialiseIncremental cborDecode)
+  where
+    -- Have to mediate between a CBOR decoder that consumes strict bytestrings
+    -- and our choice here that consumes lazy bytestrings.
+    go :: [BS.ByteString] -> CBOR.IDecode s a
+       -> m (DecodeStep LBS.ByteString CBOR.DeserialiseFailure m a)
+    go _  (CBOR.Done  trailing _ x)
+      | BS.null trailing    = return (DecodeDone x Nothing)
+      | otherwise           = return (DecodeDone x (Just trailing'))
+                                where trailing' = LBS.fromStrict trailing
+    go _  (CBOR.Fail _ _ e) = return (DecodeFail e)
+
+    -- We keep a bunch of chunks and supply the CBOR decoder with them
+    -- until we run out, when we go get another bunch.
+    go (c:cs) (CBOR.Partial  k) = go cs =<< liftST (k (Just c))
+    go []     (CBOR.Partial  k) = return $ DecodePartial $ \mbs -> case mbs of
+                                    Nothing -> go [] =<< liftST (k Nothing)
+                                    Just bs -> go cs (CBOR.Partial k)
+                                      where cs = LBS.toChunks bs
```

Comment:
That's nice, avoiding evaluation of lazy chunk which `toStrict` would do.

Reply:
And it avoids copying/concatenating them.